File: spring-aws-s3-java/src/main/java/com/migangqui/spring/aws/s3/config/AmazonS3Config.java
package com.migangqui.spring.aws.s3.config;

import com.amazonaws.auth.AWSCredentials;
import com.amazonaws.auth.AWSCredentialsProvider;
import com.amazonaws.auth.BasicAWSCredentials;
import static com.amazonaws.client.builder.AwsClientBuilder.EndpointConfiguration;
import com.amazonaws.services.s3.AmazonS3;
import com.amazonaws.services.s3.AmazonS3ClientBuilder;
import com.migangqui.spring.aws.s3.property.AmazonS3Properties;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

@Slf4j
@Configuration
public class AmazonS3Config {

    @Autowired
    private AmazonS3Properties properties;

    @Bean
    public AmazonS3 amazonS3Client() {
        AmazonS3 client;

        if (properties.isLocalstackEnabled()) {
            log.info("Registering AmazonS3Client (with Localstack)");
            client = AmazonS3ClientBuilder.standard()
                    .withEndpointConfiguration(new EndpointConfiguration(properties.getLocalstackEndpoint(), properties.getLocalstackRegion()))
                    .withPathStyleAccessEnabled(true)
                    .build();
        } else {
            log.info("Registering AmazonS3Client");
            client = AmazonS3ClientBuilder.standard()
                    .withCredentials(awsS3CredentialsProvider())
                    .withRegion(properties.getRegion())
                    .build();
        }
        return client;
    }

    private AWSCredentialsProvider awsS3CredentialsProvider() {
        return new AWSCredentialsProvider() {

            public AWSCredentials getCredentials() {
                return new BasicAWSCredentials(
                        properties.getS3AccessKey(), properties.getS3SecretKey());
            }

            public void refresh() {
                // Not implemented cause not necesary
            }
        };
    }

}

File: spring-aws-s3-kotlin/src/main/kotlin/com/migangqui/spring/aws/s3/config/AmazonS3Config.kt
package com.migangqui.spring.aws.s3.config

import com.amazonaws.auth.AWSCredentials
import com.amazonaws.auth.AWSCredentialsProvider
import com.amazonaws.auth.BasicAWSCredentials
import com.amazonaws.client.builder.AwsClientBuilder.EndpointConfiguration
import com.amazonaws.services.s3.AmazonS3
import com.amazonaws.services.s3.AmazonS3ClientBuilder
import com.migangqui.spring.aws.s3.property.AmazonS3Properties
import mu.KotlinLogging
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration

@Configuration
class AmazonS3Config(private val properties: AmazonS3Properties) {

    private val log = KotlinLogging.logger {}

    @Bean
    fun amazonS3Client(): AmazonS3 {
        return if (properties.isLocalstackEnabled) {
            log.info("Registering AmazonS3Client (with Localstack)")
            AmazonS3ClientBuilder.standard()
                    .withEndpointConfiguration(EndpointConfiguration(properties.localstackEndpoint, properties.localstackRegion))
                    .withPathStyleAccessEnabled(true)
                    .build()
        } else {
            log.info("Registering AmazonS3Client")
            AmazonS3ClientBuilder.standard()
                    .withCredentials(awsS3CredentialsProvider())
                    .withRegion(properties.region)
                    .build()
        }
    }

    private fun awsS3CredentialsProvider(): AWSCredentialsProvider {
        return object : AWSCredentialsProvider {
            override fun getCredentials(): AWSCredentials {
                return BasicAWSCredentials(
                        properties.s3AccessKey, properties.s3SecretKey)
            }

            override fun refresh() {
                // Not implemented cause not necesary
            }
        }
    }

}